node/core/pvf/src/prepare/pool.rs

```diff
@@ -0,0 +1,336 @@
+	to_pool: mpsc::Receiver<ToPool>,
+	from_pool: mpsc::UnboundedSender<FromPool>,
+	spawned: HopSlotMap<Worker, WorkerData>,
+	mux: Mux,
+}
+
+/// A fatal error that warrants stopping the event loop of the pool.
+struct Fatal;
+
+async fn run(
+	Pool {
+		program_path,
+		spawn_timeout,
+		to_pool,
+		mut from_pool,
+		mut spawned,
+		mut mux,
+	}: Pool,
+) {
+	macro_rules! break_if_fatal {
+		($expr:expr) => {
+			match $expr {
+				Err(Fatal) => break,
+				Ok(v) => v,
+			}
+		};
+	}
+
+	let mut to_pool = to_pool.fuse();
+
+	loop {
+		futures::select! {
+			to_pool = to_pool.next() => {
+				let to_pool = break_if_fatal!(to_pool.ok_or(Fatal));
+				handle_to_pool(
+					&program_path,
+					spawn_timeout,
+					&mut spawned,
+					&mut mux,
+					to_pool,
+				)
+			}
+			ev = mux.select_next_some() => break_if_fatal!(handle_mux(&mut from_pool, &mut spawned, ev)),
+		}
+
+		break_if_fatal!(purge_dead(&mut from_pool, &mut spawned).await);
+	}
+}
+
+async fn purge_dead(
+	from_pool: &mut mpsc::UnboundedSender<FromPool>,
+	spawned: &mut HopSlotMap<Worker, WorkerData>,
+) -> Result<(), Fatal> {
+	let mut to_remove = vec![];
+	for (worker, data) in spawned.iter_mut() {
+		if data.idle.is_none() {
+			// The idle token is missing, meaning this worker is now occupied: skip it. This is
+			// because the worker process is observed by the work task and should it reach the
+			// deadline or be terminated it will be handled by the corresponding mux event.
+			continue;
+		}
+
+		if let Poll::Ready(()) = futures::poll!(&mut data.handle) {
+			// a resolved future means that the worker has terminated. Weed it out.
+			to_remove.push(worker);
+		}
+	}
+	for w in to_remove {
+		let _ = spawned.remove(w);
+		reply(from_pool, FromPool::Rip(w))?;
+	}
+	Ok(())
+}
+
+fn handle_to_pool(
+	program_path: &Path,
+	spawn_timeout: Duration,
+	spawned: &mut HopSlotMap<Worker, WorkerData>,
+	mux: &mut Mux,
+	to_pool: ToPool,
+) {
+	match to_pool {
+		ToPool::Spawn => {
+			mux.push(spawn_worker_task(program_path.to_owned(), spawn_timeout).boxed());
+		}
+		ToPool::StartWork {
+			worker,
+			code,
+			artifact_path,
+			background_priority,
+		} => {
+			if let Some(data) = spawned.get_mut(worker) {
+				if let Some(idle) = data.idle.take() {
+					mux.push(
+						start_work_task(worker, idle, code, artifact_path, background_priority)
+							.boxed(),
+					);
+				} else {
+					// idle token is present after spawn and after a job is concluded;
+					// the precondition for `StartWork` is it should be sent only if all previous work
+					// items concluded;
+					// thus idle token is Some;
+					// qed.
+					never!();
+				}
+			} else {
+				// That's a relatively normal situation since the queue may send `start_work` and
+				// before receiving it the pool would report that the worker died.
+			}
+		}
+		ToPool::Kill(worker) => {
+			// It may be absent if it were previously already removed by `purge_dead`.
+			let _ = spawned.remove(worker);
```

Comment:
How does that kill the worker?

Reply:
Through `Drop`. The value type is `WorkerData` which owns a `worker_common::WorkerHandle` which owns `async_process::Child`, which is configured to kill the process on drop.